drivers/timer/mcux_os_timer.c

```diff
@@ -143,23 +149,26 @@ static uint32_t mcux_lpc_ostick_compensate_system_timer(void)
 		uint32_t slept_time_ticks;
 		uint32_t slept_time_us;
 
-		counter_stop(counter_dev);
-
-		counter_get_value(counter_dev, &slept_time_ticks);
-
-		if (!(counter_is_counting_up(counter_dev))) {
-			slept_time_ticks = counter_get_top_value(counter_dev) - slept_time_ticks;
-		}
-		slept_time_us = counter_ticks_to_us(counter_dev, slept_time_ticks);
-		cyc_sys_compensated += (k_us_to_ticks_floor32(slept_time_us) * CYC_PER_TICK);
+		if (counter_running) {
+			counter_stop(counter_dev);
+			counter_running = false;
+			counter_get_value(counter_dev, &slept_time_ticks);
 
+			if (!(counter_is_counting_up(counter_dev))) {
+				slept_time_ticks = counter_get_top_value(counter_dev) -
+						   slept_time_ticks;
+			}
+			slept_time_us = counter_ticks_to_us(counter_dev, slept_time_ticks);
+			cyc_sys_compensated += CYC_PER_US * slept_time_us;
 #if CONFIG_MCUX_OS_TIMER_PM_POWERED_OFF
-		/* Reactivate os_timer for cases where it loses its state */
-		OSTIMER_Init(base);
+			/* Reset the OS Timer to a known state */
+			RESET_PeripheralReset(kOSEVENT_TIMER_RST_SHIFT_RSTn);
+			/* Reactivate os_timer for cases where it loses its state */
+			OSTIMER_Init(base);
 #endif
-
-		/* Announce the time slept to the kernel*/
-		mcux_lpc_ostick_isr(NULL);
+			/* Announce the time slept to the kernel*/
+			mcux_lpc_ostick_isr(NULL);
+		}
 	} else {
 		ret = 1;
 	}
```

Comment:
this whole function is confusing to read, it can be cleaned up by inverting the conditionals, like

```c
if (!counter_dev) {
   return 1;
}

if (!counter_running) {
   return 0;
}

/* rest of the code */

return 0;
```